### PR TITLE
Add explicit linking to `atomic` library when required

### PR DESCRIPTION
**Pull request by csparker247**
This PR checks if the compiler provides `std::atomic` as an external library and links PTL libraries against it if required. This is essentially an update of @LocutusOfBorg's changes for PTLv3.

### CODE REVIEW

**Comment by csparker247** [comment]
I've tested that this still builds fine on my M1 MBP. 

I also want to draw attention here to [this comment](https://github.com/tomopy/tomopy/issues/628#issuecomment-1782494105). In short, the original intent of this was to support riscv64 on gcc when gcc didn't provide atomic natively. Since the original contribution, gcc has added native support for atomic on riscv64, so these changes aren't strictly necessary for that platform, but may be useful for others.

**Comment by csparker247** [comment]
I also meant to ask, do you want to open a PR to the upstream from `tomopy/PTL`, or should I just go ahead and open it myself?

File: cmake/Modules/CheckAtomic.cmake [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+    uint64_t i = x.load(std::memory_order_relaxed);
+    return 0;
+  }
+  " ${varname})
+  set(CMAKE_REQUIRED_FLAGS ${OLD_CMAKE_REQUIRED_FLAGS})
+endfunction()
+
+# Check for (non-64-bit) atomic operations.
+if(MSVC OR APPLE) # apple do not need latomic link
+  set(HAVE_CXX_ATOMICS_WITHOUT_LIB True)
```
Comment (carterbox) [inline cmake/Modules/CheckAtomic.cmake]: Why do we just skip directly to setting this variable for MSVC and APPLE? Isn't the point of this test to actually check if the code snips above compile?

**Review by carterbox — CHANGES_REQUESTED** [review]
Thanks for opening this PR. I think these changes are still necessary if downstream users only have access to version of GCC before they added atomics support.

Please address linting errors from the upstream CI checks that I just activated. I'm not sure that the other fails are related to this PR.

File: cmake/Modules/CheckAtomic.cmake (hunk 1 above)
Comment (csparker247) [reply cmake/Modules/CheckAtomic.cmake]: That's a good question. It was part of the original commits, and I can at least confirm that removing `APPLE` there breaks my build. But really the compiler check should handle that, I would think. I'll look closer at the rest of the module logic.

Comment (csparker247) [reply cmake/Modules/CheckAtomic.cmake]: ~~I need to lint my latest changes, but~~ I've refactored CheckAtomic in a way that should be more generic across platforms.

File: cmake/Modules/CheckAtomic.cmake [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+needed for atomic support. If it is needed libatomicis added to
+``CMAKE_REQUIRED_LIBRARIES``. So after running CheckAtomic you can use
+std:atomic.
+
+Since 5.75.0.
+#]=======================================================================]
+
+include(CMakePushCheckState)
+include(CheckCXXSourceCompiles)
+include(CheckLibraryExists)
```
Comment (csparker247) [inline cmake/Modules/CheckAtomic.cmake]:
```suggestion
```

**Pull request merged**